### ISSUE TITLE
Windows stack canaries obscure dataflow

### ISSUE THREAD

**Issue by yrp604** [Vector35/binaryninja-api#2250]
Here is a simple C program that forces the use of canaries in `quux`
```
size_t quux(const char *p) {
    char buf[0x10] = { 0 };
    strcpy(buf, p);

    return strlen(buf); // NOTE: we return a strlen()
}

int main(int argc, char **argv) {
    printf("%d\n", quux(argv[0]));
}
```

If we compile this as an x64 PE and load this into binja and fix up the names, we get the following:
```
int64_t quux(char* arg1)
140001011  void var_48
140001011  int64_t rax_1 = *data_14001d018 ^ &var_48
14000101e  void var_28
14000101e  void* rdi = &var_28
140001028  for (int64_t rcx = 0x10; rcx != 0; rcx = rcx - 1)
140001028      *rdi = 0
140001028      rdi = rdi + 1
140001034  strcpy(&var_28, arg1)
14000103e  strlen(&var_28)
140001057  return canary_check(rax_1 ^ &var_48)

int64_t main(int32_t arg1, void* arg2)
14000108e  printf(data_14001d000, quux(*(arg2 + 0)))
140001099  return 0
```

Note that in quux, at 103e we compute the strlen, and according to hlil, discard it to return the value of the cookie check function. This is obviously inaccurate.

If we look at the asm for that canary_check function, it has a custom calling convention that does not touch rax:
```
canary_check:
140001130  483b0de1be0100     cmp     rcx, qword [rel data_14001d018]
140001137  f27512             bnd jne 0x14000114c  {data_14001d018}

14000113a  48c1c110           rol     rcx, 0x10
14000113e  66f7c1ffff         test    cx, 0xffff
140001143  f27502             bnd jne 0x140001148

140001146  f2c3               bnd retn  {__return_addr}

140001148  48c1c910           ror     rcx, 0x10

14000114c  e9ab020000         jmp     sub_1400013fc
```

This problem is even worse in the the arm64 case -- because the canary is installed by a function that also has a custom calling convention, dataflow in _and_ out of the function is lost:

```
int64_t quux()
140001e94  int64_t var_20 = 0
140001e94  int64_t var_18 = 0
140001ea0  strcpy(&var_20, canary_install())
140001ec4  return canary_check(sx.q(strlen(&var_20)))
```

Note the lack of function arguments, because the canary_install function clobbers x0. canary_install is implemented as :
```
canary_install:
140001000  ff4300d1   sub     sp, sp, #0x10
140001004  f1000090   adrp    x17, 0x14001d000
140001008  310a40f9   ldr     x17, [x17, #0x10]  {data_14001d010}
14000100c  f16331cb   sub     x17, sp, x17 {var_10}
140001010  f10700f9   str     x17, [sp, #0x8 {var_8}]
140001014  c0035fd6   ret     
```

While I'm capable of fixing these functions as they pop up, this seems to be a use-case that should be well supported.

I see two possible fixes, though obviously I'm open to others:
1. Introduce a naked or empty calling convention -- ideally these functions would end up tagged as such and thus not affect the arguments in/out of the functions. I don't have a good sense for other bad side effects this might have, so it's entirely possible its a bad idea.
2. Platform specific heuristics/fixups -- identify the stack check functions on WIndows PE's and fiddle with the calling convention only on these functions to correct dataflow.

**Comment by yrp604** [comment]
Followup, there are two distinct things happening here.

1. Input arguments to a function being clobbered when they shouldn't be (the arm64 input case). From talking to Ryan this is due to the stack pointer being adjusted inside the canary install function.

2. The canary_check function on both platforms is basically this:
```
def canary_check():
  if canary is good:
      return // without clobbering any registers
  else:
    kill()

def kill():
  TerminateProcess(GetCurrentProcess())
```

`kill` is noreturn, however binja cannot determine this with some further enlightenment/heuristics. As a result, kill is considered returnable, and therefor the `canary_check` function ends up clobbering all the registers that `kill` clobbers (which includes `rax`). With kill updated to be noreturn, analysis works as expected.

The easiest solution might just be to have a windows.py or something that looks for TerminateProcess called on the result of GetCurrentProcess and flag those functions as noreturn? This pattern would also apply to RaiseException with certain exception codes.

**Comment by CouleeApps** [comment]
The bug specifically in this issue is due to noreturn analysis taking precedence over typelib noreturn annotations. Will be fixed in 3.1.3484-dev
